File: hard/Lc315/Lc315.py
# 归并排序
# NlogN
#
class Struct:
    def __init__(self, num, i) -> None:
        self.num = num
        self.i = i

class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        newNums = []
        ans = [0]*len(nums)

        for i, num in enumerate(nums):
            newNums.append(Struct(num, i))

        def mergeSort(nums, left, right):
            if left >= right:
                return 
            mid = (left+right+1)//2
            mergeSort(nums, left, mid-1)
            mergeSort(nums, mid, right)
            merge(nums, left, mid, right)

        def merge(nums, left, mid, right):
            res = []
            i, j = left, mid
            while i <= mid-1 and j <= right:
                if nums[i].num <= nums[j].num:
                    res.append(nums[i])
                    ans[nums[i].i] += j-mid+1
                    i += 1
                else:
                    res.append(nums[j])
                    j += 1
            res += nums[i:mid]
            res += nums[j:right+1]
            for k in range(i,mid):
                ans[nums[k].i] += right-mid+1
            nums[left:right+1] = res

        mergeSort(newNums, 0, len(newNums)-1)
        return ans

File: hard/Lc315/Lc315.go
package Lc315

// [5,2,6,1]
// 比i小的第一个元素的位置
// f(i)
// 归并排序
// a---b---c
// 递归性质
//
func countSmaller(nums []int) []int {
	ans := make([]int, len(nums))
	var mergeSort func(numIndexs []numIndex, left int, right int)
	mergeSort = func(numIndexs []numIndex, left int, right int) {
		if left == right {
			return
		}
		mid := (left + right + 1) / 2
		mergeSort(numIndexs, left, mid-1)
		mergeSort(numIndexs, mid, right)
		//merge
		merged := make([]numIndex, 0, right-left+1)
		i, j := left, mid
		for i < mid && j <= right {
			if numIndexs[i].num > numIndexs[j].num {
				merged = append(merged, numIndexs[i])
				ans[numIndexs[i].index] += right - j + 1
				i += 1
			} else {
				merged = append(merged, numIndexs[j])
				j += 1
			}
		}
		for k := i; k < mid; k++ {
			merged = append(merged, numIndexs[k])
		}
		for k := j; k <= right; k++ {
			merged = append(merged, numIndexs[k])
		}
		for k := 0; k < len(merged); k++ {
			numIndexs[left+k] = merged[k]
		}
	}
	numIndexs := make([]numIndex, len(nums))
	for i, num := range nums {
		numIndexs[i] = numIndex{num, i}
	}
	mergeSort(numIndexs, 0, len(numIndexs)-1)
	return ans
}

type numIndex struct {
	num   int
	index int
}
